### ISSUE TITLE
Error: "different number of segments" – EraseOpenCorners ufo2ft filter breaks compatibility in some glyphs

### ISSUE THREAD

**Issue by arrowtype** [googlefonts/fontmake#897]
This isn’t necessarily an issue in FontMake, and maybe it’s not an issue at all ... but it threw me for about an hour of debugging in a variable font build, so I wanted to record it here to point to a potential problem, and to perhaps help the next person who might stumble into this same obstacle.

I had a collection of sources that I knew were compatible to a high degree of certainty ... at least, I was checking them in every way I knew how to, after many previous variable font builds using FontMake. Turns out, I found another thing to check for in the future!

So, here was my problem: I was getting this error on a few glyphs within my designspace:

```
ERROR:cu2qu.ufo:Glyphs named 'Emcyr' have different number of segments
```

...and yet, I triple-checked that these glyphs had the same number of segments, the exact same start points, the same number of points, and the same types of segments.

I was stumped until I read through the FontMake output again, more closely. I spotted a line that was new to me:

> `INFO:ufo2ft.filters.base:Running EraseOpenCornersFilter on ShantellSans-ExtraBoldItalic`

Well, a primary change between this build and the last one is that I had run the sources through Glyphs, for kerning, then output back to UFOs, using ufo2glyphs then glyphs2ufo, respectively. 

I know that handling open corners in builds has been a long goal in FontMake builds, and it looks like it’s finally happening! That’s great news overall, but I realize it must be very difficult to detect open corners with 100% accuracy, so some glyphs were tripping it up, in some of my sources.

To prevent this behavior, I had to add the following lib key to my source UFOs:

```
    <key>com.github.googlei18n.ufo2ft.filters</key>
    <array>
      <dict>
        <key>name</key>
        <string>EraseOpenCorners</string>
        <key>pre</key>
        <integer>0</integer>
      </dict>
    </array>
```

(This is next to a couple of other important ufo2ft filter keys, `decomposeTransformedComponents` and `flattenComponents`, which are both set with `pre: 1`). 

## My questions

(I’ll try to answer these, but I want to submit the issue first).

1. Was `EraseOpenCorners` recently added to FontMake or ufo2ft as a default step, or was this only activated by running things through GlyphsApp or GlyphsLib?
2. If this was triggered by going through GlyphsApp/GlyphsLib, where is that trigger stored in the UFO? I’m having a bit of trouble finding it when I search, but maybe I just need to keep looking.
3. Might there be any opportunity to improve the algorithm for the `EraseOpenCornersFilter`, or is it just a matter of it being impossible for a computer to do this 100% perfectly? I’m happy to share my repo privately if it would help, and it is intended for public, OFL release in a month or two.

Anyway, thanks for reading, and hopefully this helps save someone else some time!

**Comment by arrowtype** [comment]
After some more searching, all I can find in the way of this filter is [glyphsLib/eraseOpenCorners.py](https://github.com/googlefonts/glyphsLib/blob/ee5e1fbbde0e40a0352c5c4444e4131020662f6a/Lib/glyphsLib/filters/eraseOpenCorners.py), but that is confusing, as the FontMake output presents this as a `ufo2ft` filter. 

Maybe there is something in the naming that I don’t quite understand? My guess is, the ufo2ft filter base is used by glyphsLib, and then FontMake communicates this as though it’s coming from ufo2ft? Is this obviously correct, or could it be clarified in the output?

I also can’t locate _why_ that would be triggered in the build, unless it’s just generally happening because there is Glyphs stuff in the font libs. If there is a way to prevent that without adding `EraseOpenCorners = {pre:0}` to my font lib, it would be handy to know about for the future!

Thanks again, and thanks as always for building and maintaining such a valuable tool!

**Comment by anthrotype** [comment]
> Was EraseOpenCorners recently added to FontMake or ufo2ft as a default step, or was this only activated by running things through GlyphsApp or GlyphsLib?

automatically, only when going through glyphsLib, since it's a GlyphsApp's only behaviour (and IMHO very idiosyncratic)

> where is that trigger stored in the UFO? 

exactly where you found it, in the `com.github.googlei18n.ufo2ft.filters` key of the UFO lib.plist

> improve the algorithm for the EraseOpenCornersFilter

since it's a glyphsLib feature (implemented as a ufo2ft custom plug-in filter), I suggest to file an issue over at glyphsLib

> that is confusing, as the FontMake output presents this as a ufo2ft filter.

maybe it is. glyphsLib defines the filter module, but it's ufo2ft which actuallt runs the filter, and that's where the logging message is emitted.

Note that "pre=False" simply means that the filter will be run as a _post_-filter, i.e. after the default filters (decomposing components, removing overlaps, cu2qu, etc.). It doesn't mean that it will not be run! Apologies if that confuses you. The terminology of (post)Filter vs preFilter comes from Glyphs.app itself..

So basically, yes, when building from .glyphs sources, fontmake uses glyphsLib to convert them to DS+UFOs and glyphsLib automatically adds that filter. I was hesitant to add it by default to all builds, but people argued that since this is done automatically by default in Glyphs.app, glyphsLib should do the same.  
In your case, I think the filter sneaked in when you went ufo2glyphs and then back glyphs2ufo (specifically in the latter step).

Ok now to the question, how does one opt out of that EraseOpenCorner filter that glyphsLib adds by default?
Let me find out for you.. 😅

**Comment by anthrotype** [comment]
well, apparently you can't! https://github.com/googlefonts/glyphsLib/issues/788 😄 
So we need a flag to glyphs2ufo to opt-out from adding that filter. PRs welcome

in the meantime, I suggest you simply edit the lib.plist and remove the entry corresponding to the unwanted filter.

**Comment by anthrotype** [comment]
hm that is weird. Are you 100% sure? I just tried with a UFO that didn't have any ufo2ft filters in the lib.plist, converted it to glyphs with ufo2glyphs and then back to UFO with ufo2glyphs and I am definitely seeing the added entry in the lib's `<dict>`:

```xml
    <key>com.github.googlei18n.ufo2ft.filters</key>
    <array>
      <dict>
        <key>name</key>
        <string>eraseOpenCorners</string>
        <key>namespace</key>
        <string>glyphsLib.filters</string>
        <key>pre</key>
        <true/>
      </dict>
    </array>
```

the fact that it kinda works for you if you run it after the default filters (as opposed to before as in pre=True) is probably coincidental. I would not rely on that as a general workaround.

**Comment by anthrotype** [comment]
BTW, the filter entry you added is missing a namespace key, which means ufo2ft thinks that is a built-in filter and attempts to import it from its own ufo2ft.filters package. But that filter is only available in the glyphsLib package, it's external to ufo2ft, it uses ufo2ft's API but doesn't belong to ufo2ft proper.

**Comment by arrowtype** [comment]
>> This is odd... that lib key wasn't in my ufo lib.plist until I specifically added it.
> hm that is weird. Are you 100% sure?

Okay, I’m back at my computer today and have been digging into this again.

Turns out, something else in my pre-build chain was causing this filter to be added. I’m not entirely sure what it was. But, if I add a step to set the ufo2ft filters as I want them set, things build as needed, without the `EraseOpenCorners` disruption.

Thanks for prompting me to look a little more closely! It’s good to know this is a solvable issue on my end.

**Comment by simoncozens** [comment]
There is still a problem here. Noto Looped Thai won't build as a variable font because erasing the open corners leaves the outlines incompatible in a number of glyphs. I think it happens because the corner "covers" a number of segments:

<img width="527" alt="Screenshot 2022-08-22 at 08 41 43" src="https://user-images.githubusercontent.com/106728/185866485-61f4a229-316e-4767-af46-93cd9260f4a7.png">

I would like to be able to turn the filter off while processing variable fonts.

**Comment by anthrotype** [comment]
> turn the filter off while processing variable fonts.

it also depends on what Glyphs.app currently does and thus what Glyphs.app users expect fontmake to do with .glyphs sources; glyphsLib should try to follow that as much as it can, while also allowing to opt-out of this (which currently is impossible, see https://github.com/googlefonts/glyphsLib/issues/788)

Does Glyphs.app run the erase open corners filter on each glyph _masters_ before these get interpolated into a VF (this is what we currently do, and we inevitably incur in interpolability issues)? Or does it only run on the instances generated from the masters' interpolation (this no erase open corners for VF basically, only for statics)?

**Comment by simoncozens** [comment]
> Does Glyphs.app run the erase open corners filter on each glyph masters before these get interpolated into a VF (this is what we currently do, and we inevitably incur in interpolability issues)?

So I tried to export a VF from Glyphs, and oh look

<img width="758" alt="Screenshot 2022-08-22 at 10 44 48" src="https://user-images.githubusercontent.com/106728/185891924-bb485fb5-b016-40fc-bce2-8a241f96902f.png">

So presumably Glyphs is trying to do the same thing we do and failing for the same reasons. This is annoying because the curves *are* compatible, or would be if there was no open corner erasure.

I'm tempted to hit "remove overlaps" and be done with it. But maybe we could be better than Glyphs here...

**Comment by anthrotype** [comment]
btw, the screenshot above does not look lika an "open" corner to me, it appeas inside the "inked" glyph area so it should not be erased, am I missing something?

**Comment by simoncozens** [comment]
I just fiddled the shapes until it did export, and yes, Glyphs does perform open corner erasure here, same as we do.

**Comment by anthrotype** [comment]
> yes, Glyphs does perform open corner erasure here, same as we do.

then I completely misunderstood what "open corners" mean...

**Comment by simoncozens** [comment]
I think Glyphs doesn't check the winding direction. If this is an open corner:
<img width="442" alt="Screenshot 2022-08-22 at 11 18 14" src="https://user-images.githubusercontent.com/106728/185898395-385a9db5-4da2-429c-a461-bfe79a492423.png">
then so is this:
<img width="311" alt="Screenshot 2022-08-22 at 11 18 24" src="https://user-images.githubusercontent.com/106728/185898430-35da8fae-810d-45f0-82cf-1f35a7e71da8.png">

**Comment by anthrotype** [comment]
I don't undestand why should the latter shape in the above screenshot (with the inner self-overlapping triangle) be considered an "open corner"?
